updater/internal/install/service_windows.go

```diff
@@ -0,0 +1,276 @@
+
+	s, err := m.OpenService(w.serviceName)
+	if err != nil {
+		return fmt.Errorf("failed to open service: %w", err)
+	}
+	defer s.Close()
+
+	if _, err := s.Control(svc.Stop); err != nil {
+		return fmt.Errorf("failed to start service: %w", err)
+	}
+
+	return nil
+}
+
+// Installs the service
+func (w windowsService) Install() error {
+	wsc, err := readWindowsServiceConfig(w.newServiceFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to read service config: %w", err)
+	}
+
+	iDir, err := installDirFromRegistry(w.productName)
+	if err != nil {
+		return fmt.Errorf("failed to get install dir: %w", err)
+	}
+
+	expandArguments(wsc, w.productName, iDir)
+
+	splitArgs, err := shellquote.Split(wsc.Service.Arguments)
+	if err != nil {
+		return fmt.Errorf("failed to parse arguments in service config: %w", err)
+	}
+
+	startType, delayed, err := startType(wsc.Service.Start)
+	if err != nil {
+		return fmt.Errorf("failed to parse start type in service config: %w", err)
+	}
+
+	m, err := mgr.Connect()
+	if err != nil {
+		return fmt.Errorf("failed to connect to service manager: %w", err)
+	}
+	defer m.Disconnect()
+
+	s, err := m.CreateService(w.serviceName,
+		filepath.Join(iDir, wsc.Path),
+		mgr.Config{
+			Description:      wsc.Service.Description,
+			DisplayName:      wsc.Service.DisplayName,
+			StartType:        startType,
+			DelayedAutoStart: delayed,
+		},
+		splitArgs...,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create service: %w", err)
+	}
+	defer s.Close()
+
+	return nil
+}
+
+// Uninstalls the service
+func (w windowsService) Uninstall() error {
+	m, err := mgr.Connect()
+	if err != nil {
+		return fmt.Errorf("failed to connect to service manager: %w", err)
+	}
+	defer m.Disconnect()
+
+	s, err := m.OpenService(w.serviceName)
+	if err != nil {
+		return fmt.Errorf("failed to open service: %w", err)
+	}
+
+	if err = s.Delete(); err != nil {
+		sCloseErr := s.Close()
+		if sCloseErr != nil {
+			log.Default().Printf("Failed to close service: %s\n", err)
+		}
+		return fmt.Errorf("failed to delete service: %w", err)
+	}
+
+	if err := s.Close(); err != nil {
+		return fmt.Errorf("failed to close service: %w", err)
+	}
+
+	// Wait for the service to actually be deleted:
+	for {
+		s, err := m.OpenService(w.serviceName)
+		if err != nil {
+			if err.Error() == serviceNotExistErrStr {
+				// This is expected when the service is uninstalled.
+				break
+			}
+			return fmt.Errorf("got unexpected error when waiting for service deletion: %w", err)
+		}
+
+		if err := s.Close(); err != nil {
+			return fmt.Errorf("failed to close service: %w", err)
+		}
+		// rest with the handle closed to let the service manager remove the service
+		time.Sleep(uninstallServicePollInterval)
+	}
+	return nil
+}
+
+type windowsServiceConfig struct {
+	Path string `json:"path"`
+	// Note: Name is a part of the on disk config, but we keep the service name hardcoded; We do not want to use a different service name.
+	Service struct {
```

Comment:
Any reason we're creating an anonymous struct instead of declaring a type for this?

Reply:
Not really, just wasn't really much of a functional reason to (the struct isn't used anywhere by itself). I have broke this out into a separate named struct.